lib/utils/ignore.ts

```diff
@@ -0,0 +1,151 @@
+import * as _ from 'lodash';
+import { fs } from 'mz';
+import * as path from 'path';
+
+import dockerIgnore = require('@zeit/dockerignore');
+import ignore from 'ignore';
+
+export enum IgnoreFileType {
+	DockerIgnore,
+	GitIgnore,
+}
+
+interface IgnoreEntry {
+	pattern: string;
+	// The relative file path from the base path of the build context
+	filePath: string;
+}
+
+export class FileIgnorer {
+	private dockerIgnoreEntries: IgnoreEntry[];
+	private gitIgnoreEntries: IgnoreEntry[];
+
+	private static ignoreFiles: Array<{
+		pattern: string;
+		type: IgnoreFileType;
+		allowSubdirs: boolean;
+	}> = [
+		{
+			pattern: '.gitignore',
+			type: IgnoreFileType.GitIgnore,
+			allowSubdirs: true,
+		},
+		{
+			pattern: '.dockerignore',
+			type: IgnoreFileType.DockerIgnore,
+			allowSubdirs: false,
+		},
+	];
+
+	public constructor(public basePath: string) {
+		this.dockerIgnoreEntries = [];
+		this.gitIgnoreEntries = [];
+	}
+	/**
+	 * @param  {string} relativePath
+	 * 	The relative pathname from the build context, for example a root level .gitignore should be
+	 * 		./.gitignore
+	 * @returns IgnoreFileType
+	 * 	The type of ignore file, or null
+	 */
+	public isIgnoreFile(relativePath: string): IgnoreFileType | null {
+		for (const { pattern, type, allowSubdirs } of FileIgnorer.ignoreFiles) {
+			if (relativePath === pattern || relativePath === `./${pattern}`) {
```

Comment:
It's because of the fact that dockerignore files can only appear in the root, so if I only checked the basename, it would allow them in every directory. I also can't remove the `./${pattern}` check, because then a root level dockerignore file wouldn't be detected. Keen to hear any other ideas you have though, I could be missing something obvious.